src/core/api/dns_server_api.cpp

```diff
@@ -73,4 +75,26 @@ void otDnssdQueryHandleDiscoveredHost(otInstance *aInstance, const char *aHostFu
     instance.Get<Dns::ServiceDiscovery::Server>().HandleDiscoveredHost(aHostFullName, *aHostInfo);
 }
 
+const otDnssdQuery *otDnssdGetNextQuery(otInstance *aInstance, const otDnssdQuery *aQuery)
+{
+    Instance &              instance = *static_cast<Instance *>(aInstance);
+    const QueryTransaction *query    = static_cast<const QueryTransaction *>(aQuery);
+
+    return instance.Get<Dns::ServiceDiscovery::Server>().GetNextQuery(query);
+}
+
+otDnssdQueryType otDnssdGetQueryTypeAndName(const otDnssdQuery *aQuery, char *aNameOutput, uint16_t aNameSize)
+{
+    otDnssdQueryType        type  = OT_DNSSD_QUERY_TYPE_NONE;
+    const QueryTransaction *query = static_cast<const QueryTransaction *>(aQuery);
+
+    OT_ASSERT(aQuery != nullptr);
+    OT_ASSERT(aNameOutput != nullptr);
+    OT_ASSERT(query->IsValid());
+    type = static_cast<otDnssdQueryType>(Dns::ServiceDiscovery::Server::GetQueryTypeAndName(
+        query->GetResponseHeader(), query->GetResponseMessage(), aNameOutput, aNameSize));
```

Comment:
```suggestion
    type = static_cast<otDnssdQueryType>(Dns::ServiceDiscovery::Server::GetQueryTypeAndName(query, aNameOutput, aNameSize));
```

So that we don't need to expose `QueryTransaction` publicly.